src/atlas/interpolation/method/bil/BilinearRemapping.cc

```diff
@@ -0,0 +1,395 @@
+                const PointLonLat pll{(*olonlat_)(*i, (size_t)0), (*olonlat_)(*i, (size_t)1)};  // lookup point
+                msg << "\t(lon,lat) = " << pll << "\n";
+            }
+
+            Log::error() << msg.str() << std::endl;
+            throw_Exception(msg.str());
+        }
+    }
+
+    // fill sparse matrix and return
+    Matrix A(out_npts, inp_npts, weights_triplets);
+    matrix_shared_->swap(A);
+}
+
+Method::Triplets BilinearRemapping::projectPointToElements(size_t ip, const ElemIndex2::NodeList& elems,
+                                                           std::ostream& /* failures_log */) const {
+    ATLAS_ASSERT(elems.begin() != elems.end());
+
+    const size_t inp_points = ilonlat_->shape(0);
+    std::array<size_t, 4> idx;
+    std::array<double, 4> w;
+
+    Triplets triplets;
+    triplets.reserve(4);
+
+    double lon{(*olonlat_)(ip, 0)};
+    while (lon > 360.0) { lon -= 360.0; }
+    while (lon < 0.0) { lon += 360.0; }
+    PointXY ob_loc{lon, (*olonlat_)(ip, 1)};  // lookup point
+
+    idx_t single_point;
+    for (ElemIndex2::NodeList::const_iterator itc = elems.begin(); itc != elems.end(); ++itc) {
+        const idx_t elem_id = idx_t((*itc).value().payload());
+        ATLAS_ASSERT(elem_id < connectivity_->rows());
+
+        const idx_t nb_cols = connectivity_->cols(elem_id);
+        ATLAS_ASSERT(nb_cols == 3 || nb_cols == 4);
+
+        for (idx_t i = 0; i < nb_cols; ++i) {
+            idx[i] = (*connectivity_)(elem_id, i);
+            ATLAS_ASSERT(idx[i] < inp_points);
+        }
+
+        constexpr double tolerance = 1.e-12;
+
+        if (nb_cols == 3) {
+            /* triangle */
+            element::Triag2D triag(PointXY{(*ilonlat_)(idx[0], size_t(0)), (*ilonlat_)(idx[0], size_t(1))},
+                                   PointXY{(*ilonlat_)(idx[1], size_t(0)), (*ilonlat_)(idx[1], size_t(1))},
+                                   PointXY{(*ilonlat_)(idx[2], size_t(0)), (*ilonlat_)(idx[2], size_t(1))});
+
+            // pick an epsilon based on a characteristic length (sqrt(area))
+            // (this scales linearly so it better compares with linear weights u,v,w)
+            const double edgeEpsilon = parametricEpsilon * std::sqrt(triag.area());
+            ATLAS_ASSERT(edgeEpsilon >= 0);
+
+            Intersect is = triag.intersects(ob_loc, edgeEpsilon);
+
+            if (is) {
+                // weights are the linear Lagrange function evaluated at u,v (aka
+                // barycentric coordinates)
+                w[0] = 1. - is.u - is.v;
+                w[1] = is.u;
+                w[2] = is.v;
+
+                for (size_t i = 0; i < 3; ++i) {
+                    triplets.emplace_back(ip, idx[i], w[i]);
+                }
+
+                break;  // stop looking for elements
+            }
+        }
+        else {
+            /* quadrilateral */
+            element::Quad2D quad(PointXY{(*ilonlat_)(idx[0], (size_t)0), (*ilonlat_)(idx[0], (size_t)1)},
```

Comment:
You can omit the C-style casts here (size_t)

Reply:
thanks, I have removed in the latest commit